hudi-common/src/main/java/org/apache/hudi/common/model/HoodieAvroRecordMerger.java

```diff
@@ -43,9 +43,29 @@ public String getMergingStrategy() {
   }
 
   @Override
-  public Option<Pair<HoodieRecord, Schema>> merge(HoodieRecord older, Schema oldSchema, HoodieRecord newer, Schema newSchema, TypedProperties props) throws IOException {
-    return combineAndGetUpdateValue(older, newer, newSchema, props)
-        .map(r -> Pair.of(new HoodieAvroIndexedRecord(r), r.getSchema()));
+  public Option<Pair<HoodieRecord, Schema>> merge(Option<HoodieRecord> older,
+                                                  Schema oldSchema,
+                                                  Option<HoodieRecord> newer,
+                                                  Schema newSchema,
+                                                  TypedProperties props) throws IOException {
+    boolean isValidNew = isValid(newer);
+    boolean isValidOld = isValid(older);
+    boolean isDeleteNew = isValidNew && isDelete(newer, newSchema, props);
+
+    if (!isValidOld && !isValidNew) { // No meaningful information found.
+      return Option.empty();
+    } else if (isValidOld && !isValidNew) { // Return old record for data safety.
+      return Option.of(Pair.of(older.get(), oldSchema));
+    } else if (!isValidOld) { // Either insert or delete case, return the new record.
+      return Option.of(Pair.of(newer.get(), newSchema));
+    } else {
+      if (isDeleteNew) { // delete case
+        return Option.of(Pair.of(newer.get(), newSchema));
+      }
```

Comment:
I guess this is all the complication that @danny0405 mentioned. I think we should still check for validity before calling `combineAndGetUpdateValue`. However, we can keep the handling of indert/delete to a separate method that can be overridden. Default would be to return the new record.

IMO, one method is cleaner and easy-to-use. Also, I checked what Kafka and Flink do. They also provide one method and leave the handling of empty/null keys to the concrete implementation.
Kafka Merger - https://kafka.apache.org/35/javadoc/org/apache/kafka/streams/kstream/Merger.html
Flink InternalRowMerger - https://nightlies.apache.org/flink/flink-docs-release-1.17/api/java/org/apache/flink/table/examples/java/functions/InternalRowMergerFunction.html

@danny0405 What do you think?

Reply:
If you look at the APIs from Kafka and Flink, they either uses a T or specific `RowData` there, and I guess both of the old and new values are not nullable, but take a look of our API:

1. we make it as option so the user needs to take care whether they are empty and they even needs to disinduish UPSERT and DELETES, which is totally unnecessary.
2. We have a HoodieRecord there, but there is many specific logic for hadling deletes inside Hudi which should not explose to user totally.
3. We have many ramifications for write paths for different representations of delete `HoodieRecord`, like a HoodieRecord with D as operation, a HoodieAvroEmptyRecord, a IGNORE record, I don't think any normal user should see that complexity, even we make a default impls there, the user still needs to undertand it and have a good comprehension to use it correctly.